File: openmole/core/org.openmole.core.serializer/src/main/java/org/openmole/core/serializer/internal/Serializer.java
/*
 *  Copyright (C) 2010 reuillon
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.core.serializer.internal;

import scala.Tuple2;
import com.thoughtworks.xstream.XStream;
import com.thoughtworks.xstream.XStreamException;
import java.io.File;
import org.openmole.commons.tools.io.FileInputStream;
import org.openmole.commons.tools.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Collection;
import java.util.Map;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.commons.tools.function.IPartialFunction;
import org.openmole.commons.tools.io.IHash;
import org.openmole.core.serializer.ISerializer;
import static org.openmole.commons.tools.io.FileUtil.*;

/**
 *
 * @author reuillon
 */
public class Serializer implements ISerializer {

    final XStream xstream = new XStream();

    @Override
    public <T> T deserialize(File file) throws InternalProcessingError {
        try {
            InputStream is = new FileInputStream(file);
            try {
                return (T) this.<T>deserialize(is);
            } finally {
                is.close();
            }
        } catch (IOException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public <T> T deserialize(InputStream is) throws InternalProcessingError {
        try {
            return (T) xstream.fromXML(is);
        } catch (XStreamException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public Tuple2<Map<File, IHash>, Collection<Class>> serializeFilePathAsHashGetPluginClassAndFiles(Object obj, File dir) throws InternalProcessingError, UserBadDataError {
        try {
            SerializerWithPathHashInjectionAndPluginListing serializer = SerializerWithPathHashInjectionAndPluginListingFactory.instance.borrowObject();
            try {
                File serialized = Activator.getWorkspace().newFile();

                OutputStream os = new FileOutputStream(serialized);
                try {
                    serializer.toXMLAndListPlugableClasses(obj, os);
                } finally {
                    os.close();
                }

                File hashName = new File(dir, Activator.getHashService().computeHash(serialized).toString());
                move(serialized, hashName);

                return new Tuple2<Map<File, IHash>, Collection<Class>>(serializer.getFiles(), serializer.getClasses());
            } finally {
                SerializerWithPathHashInjectionAndPluginListingFactory.instance.returnObject(serializer);
            }
        } catch (Exception ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public Tuple2<Collection<File>, Collection<Class>> serializeGetPluginClassAndFiles(Object obj, File file) throws InternalProcessingError {
        try {
            OutputStream os = new FileOutputStream(file);
            try {
                return serializeGetPluginClassAndFiles(obj, os);
            } finally {
                os.close();
            }
        } catch (IOException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public Tuple2<Collection<File>, Collection<Class>> serializeGetPluginClassAndFiles(Object obj, OutputStream os) throws InternalProcessingError {
        try {
            SerializerWithFileAndPluginListing serializer = SerializerWithFileAndPluginListingFactory.instance.borrowObject();
            try {
                serializer.toXMLAndListPlugableClasses(obj, os);
                return new Tuple2<Collection<File>, Collection<Class>>(serializer.getFiles(), serializer.getClasses());
            } finally {
                SerializerWithFileAndPluginListingFactory.instance.returnObject(serializer);
            }
        } catch (Exception ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public <T> T deserializeReplaceFiles(File file, IPartialFunction<File, File> files) throws InternalProcessingError, InternalProcessingError, InternalProcessingError {
        try {
            InputStream is = new FileInputStream(file);
            try {
                return this.<T>deserializeReplaceFiles(is, files);
            } finally {
                is.close();
            }
        } catch (IOException ex) {
            throw new InternalProcessingError(ex);
        }

    }

    @Override
    public <T> T deserializeReplaceFiles(InputStream is, IPartialFunction<File, File> files) throws InternalProcessingError, InternalProcessingError {
        try {
            DeserializerWithFileInjectionFromFile serializer = DeserializerWithFileInjectionFromFileFactory.instance.borrowObject();
            try {

                serializer.setFiles(files);
                return serializer.<T>fromXMLInjectFiles(is);
            } finally {
                DeserializerWithFileInjectionFromFileFactory.instance.returnObject(serializer);
            }
        } catch (Exception ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public void serialize(Object obj, OutputStream os) throws InternalProcessingError {
        try {
            xstream.toXML(obj, os);
        } catch (XStreamException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public void serialize(Object obj, File file) throws InternalProcessingError {
        try {
            OutputStream os = new FileOutputStream(file);
            try {
                serialize(obj, os);
            } finally {
                os.close();
            }
        } catch (IOException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public void serializeAsHash(Object obj, File dir) throws InternalProcessingError, UserBadDataError {
        try {
            File serialized = Activator.getWorkspace().newFile();
            serialize(obj, serialized);
            File hashName = new File(dir, Activator.getHashService().computeHash(serialized).toString());
            move(serialized, hashName);
        } catch (IOException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public <T> T deserializeReplacePathHash(File file, IPartialFunction<IHash, File> files) throws InternalProcessingError {
        try {
            InputStream is = new FileInputStream(file);
            try {
                return this.<T>deserializeReplacePathHash(is, files);
            } finally {
                is.close();
            }
        } catch (IOException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Override
    public <T> T deserializeReplacePathHash(InputStream it, IPartialFunction<IHash, File> files) throws InternalProcessingError {
        try {
            DeserializerWithFileInjectionFromPathHash deserializer = DeserializerWithFileInjectionFromPathHashFactory.instance.borrowObject();
            try {
                deserializer.setFiles(files);
                return deserializer.<T>fromXMLInjectFiles(it);
            } finally {
                DeserializerWithFileInjectionFromPathHashFactory.instance.returnObject(deserializer);
            }
        } catch (Exception ex) {
            throw new InternalProcessingError(ex);
        }
    }
}

File: openmole/core/org.openmole.core.serializer/src/main/scala/org/openmole/core/serializer/internal/Serializer.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.serializer.internal

import com.thoughtworks.xstream.XStream
import com.thoughtworks.xstream.XStreamException
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import org.openmole.commons.exception.InternalProcessingError
import org.openmole.core.serializer.FileInfoHash
import org.openmole.commons.tools.io.FileUtil._

import org.openmole.core.serializer.ISerializer
import Activator._

class Serializer extends ISerializer {

  val xstream = new XStream

  override def deserialize[T](file: File): T = {
    val is = new FileInputStream(file)
    try {
      return deserialize(is)
    } finally {
      is.close
    }
  }

  override def deserialize[T](is: InputStream): T = {
    xstream.fromXML(is).asInstanceOf[T]
  }

  override def serializeFilePathAsHashGetPluginClassAndFiles(obj: Object, dir: File): (Iterable[(File, FileInfoHash)], Iterable[Class[_]], File) = {
    val serializer = SerializerWithPathHashInjectionAndPluginListingFactory.borrowObject
    try {
      val serialized = workspace.newFile

      val os = new FileOutputStream(serialized)
      try {
        serializer.toXMLAndListPlugableClasses(obj, os)
      } finally {
        os.close
      }

      val hashName = new File(dir, hashService.computeHash(serialized).toString)
      if(!hashName.exists) move(serialized, hashName)

      (serializer.files, serializer.classes, hashName)
    } finally {
      SerializerWithPathHashInjectionAndPluginListingFactory.returnObject(serializer)
    }
  }

  override def serializeGetPluginClassAndFiles(obj: Object, file: File): (Iterable[File], Iterable[Class[_]]) = {
    val os = new FileOutputStream(file)
    try {
      return serializeGetPluginClassAndFiles(obj, os);
    } finally {
      os.close
    }
  }

  override def serializeGetPluginClassAndFiles(obj: Object, os: OutputStream): (Iterable[File], Iterable[Class[_]]) = {
    val serializer = SerializerWithFileAndPluginListingFactory.borrowObject
    try {
      serializer.toXMLAndListPlugableClasses(obj, os)
      (serializer.files, serializer.classes)
    } finally {
      SerializerWithFileAndPluginListingFactory.returnObject(serializer)
    }
  }

  override def deserializeReplaceFiles[T](file: File, files: PartialFunction[File, File] ): T = {
    val is = new FileInputStream(file)
    try {
      deserializeReplaceFiles[T](is, files)
    } finally {
      is.close
    }
  }

  override def deserializeReplaceFiles[T](is: InputStream, files: PartialFunction[File, File]): T = {
    val serializer = DeserializerWithFileInjectionFromFileFactory.borrowObject
    try {
      serializer.files = files
      serializer.fromXMLInjectFiles[T](is)
    } finally {
      DeserializerWithFileInjectionFromFileFactory.returnObject(serializer)
    }
  }

  override def serialize(obj: Object, os: OutputStream) = {
    xstream.toXML(obj, os)
  }

  override def serialize(obj: Object, file: File): Unit = {
    val os = new FileOutputStream(file);
    try {
      serialize(obj, os)
    } finally {
      os.close
    }
  }

  override def serializeAsHash(obj: Object, dir: File) = {
    val serialized = workspace.newFile
    serialize(obj, serialized)
    val hashName = new File(dir, hashService.computeHash(serialized).toString)
    move(serialized, hashName)
  }

  override def deserializeReplacePathHash[T](file: File, files: PartialFunction[FileInfoHash, File]): T = {
    val is = new FileInputStream(file)
    try {
      deserializeReplacePathHash[T](is, files)
    } finally {
      is.close
    }
  }

  override def deserializeReplacePathHash[T](it: InputStream, files: PartialFunction[FileInfoHash, File]) = {
    val deserializer = DeserializerWithFileInjectionFromPathHashFactory.borrowObject
    try {
      deserializer.files = files
      deserializer.fromXMLInjectFiles[T](it)
    } finally {
      DeserializerWithFileInjectionFromPathHashFactory.returnObject(deserializer)
    }
  }
}
